UPGRADE-GUIDE.md

```diff
@@ -0,0 +1,261 @@
+# Rack 3 Upgrade Guide
+
+This document is a work in progress, but outlines some of the key changes in
+Rack 3 which you should be aware of in order to update your server, middleware
+and/or applications.
+
+## Interface Changes
+
+### `config.ru` `Rack::Builder#run` now accepts block
+
+Previously, `Rack::Builder#run` method would only accept a callable argument:
+
+```ruby
+run lambda{|env| [200, {}, ["Hello World"]]}
+```
+
+This can be rewritten more simply:
+
+```ruby
+run do |env|
+  [200, {}, ["Hello World"]]
+end
+```
+
+### Response bodies can be used for bi-directional streaming
+
+Previously, the `rack.hijack` response header could be used for implementing
+bi-directional streaming (e.g. WebSockets).
+
+```ruby
+def call(env)
+  stream_callback = proc do |stream|
+    stream.read(...)
+    stream.write(...)
+  ensure
+    stream.close(...)
+  end
+
+  return [200, {'rack.hijack' => stream_callback}, []]
+end
+```
+
+This feature was optional and tricky to use correctly. You can now achieve the
+same thing by giving `stream_callback` as the response body:
+
+```ruby
+def call(env)
+  stream_callback = proc do |stream|
+    stream.read(...)
+    stream.write(...)
+  ensure
+    stream.close(...)
+  end
+
+  return [200, {}, stream_callback]
+end
+```
+
+### `Rack::Session` is now moved to an external gem
+
+Previously, `Rack::Session` was part of the `rack` gem. Not every application
+needs it, and it increases the security surface area of the `rack`, so it was
+decided to extract it into its own gem `rack-session` which can be updated
+independently.
+
+Applications that make use of `rack-session` will need to add that gem as a
+dependency:
+
+```ruby
+gem 'rack-session'
+```
+
+
+
+## Request Changes
+
+### `rack.version` is no longer required
+
+Previously, the "rack protocol version" was available in `rack.version` but it
+was not practically useful, so it has been removed as a requirement.
+
+### `rack.multithread`/`rack.multiprocess`/`rack.run_once` are no longer required
+
+Previously, servers tried to provide these keys to reflect the execution
+environment. These come too late to be useful, so they have been removed as  a
+requirement.
+
+### `rack.hijack?` now only applies to partial hijack
+
+Previously, both full and partial hijiack were controlled by the presence and
+value of `rack.hijack?`. Now, it only applies to partial hijack (which itself
+has been effectively replaced by streaming bodies).
```

Comment:
`s/itself has been effectively/now can be/`

Reply:
Done, thanks!